src/Aws/Kinesis/Client/Producer.hs

```diff
@@ -354,111 +319,95 @@ generatePartitionKey gen =
   Kin.partitionKey (T.pack name)
     & either (error ∘ T.unpack) id
 
--- | This will take up to @n@ items from a 'TBMQueue'.
+-- | This will take up to @n@ items from a 'TBMQueue', with a timeout.
 --
-takeTBMQueue
-  ∷ Int
+takeTBMQueueWithTimeout
+  ∷ ∀ α
+  . Int -- ^ number of items to get
+  → Int -- ^ timeout in microseconds
   → TBMQueue α
-  → STM [α]
-takeTBMQueue n q
-  | n <= 0 = return []
-  | otherwise = do
-      res ← tryReadTBMQueue q
-      case res of
-        Just (Just x) → (x:) <$> takeTBMQueue (n - 1) q
-        _ → return []
+  → IO [α]
+takeTBMQueueWithTimeout n timeout q = do
+  timedOutVar ← registerDelay timeout
+
+  let
+    go xs
+      | length xs ≥ n = return xs
+      | otherwise = do
+          res ← Left <$> readTBMQueue q <|> Right <$> (readTVar timedOutVar ≫= check)
+          case res of
+            Left mx → maybe (return xs) (go ∘ (:xs)) mx
+            Right () → return xs
```

Comment:
I wonder if something like the following would be easier to read (and be potentially more efficient because it avoids packing and unpacking with `Either`, though that's probably optimized away anyways).

```haskell
let 
    readItems xs = readTBMQueue q >>= \case
        Nothing -> return xs
        Just x -> go (x:xs)
    timeOut = readTVar timeOutVar >>= check
go xs = readItems xs <|> xs <$ timeOut
```